examples/graceful_setup_fail/main_test.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package setupfail
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/e2e-framework/pkg/env"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/features"
+)
+
+var testEnv env.Environment
+
+func TestMain(m *testing.M) {
```

Comment:
Hi @codegold79 
TestMain is usually reserved for setting up test resources, not running tests itself.

You probably can make your test simpler by moving your test environment into a proper TestXXX function.
There is nothing special about the testEnv variable, you can declare within a test and add step to it with no problem.

Check out these tests here - https://github.com/kubernetes-sigs/e2e-framework/blob/main/pkg/env/env_test.go

Reply:
Thanks @vladimirvivien for taking the time to review and provide feedback.
> TestMain is usually reserved for setting up test resources, not running tests itself...
> You probably can make your test simpler by moving your test environment into a proper TestXXX function.

I might be missing or misunderstanding something, but I wasn't able to figure out how to put the tests into a TestXXX function. At first, I tried to add a test to `func TestEnv_APIMethods(t *testing.T)` at https://github.com/kubernetes-sigs/e2e-framework/blob/332536f9ed7ed22fd069c5622b057a91ceeb1208/pkg/env/env_test.go#L46C6-L46C24. I also tried to add a test to `func TestEnv_Test(t *testing.T)` at https://github.com/kubernetes-sigs/e2e-framework/blob/332536f9ed7ed22fd069c5622b057a91ceeb1208/pkg/env/env_test.go#L137C1-L137C32.

With both TestXXX functions, I had the same problem: the [defer block](https://github.com/kubernetes-sigs/e2e-framework/blob/332536f9ed7ed22fd069c5622b057a91ceeb1208/pkg/env/env.go#L353-L377) is not reached in `func (e *testEnv) Run(m *testing.M) (exitCode int)` at https://github.com/kubernetes-sigs/e2e-framework/blob/332536f9ed7ed22fd069c5622b057a91ceeb1208/pkg/env/env.go#L345C1-L345C52.

I believe to reach that block, my only option is to call `func (e *testEnv) Run(m *testing.M) (exitCode int)`. But that funciton requires passing in `m *testing.M`. And the only way I know to pass that variable in is via `TestMain(...)`. Hench why I added an example package to do the testing, as all the examples use TestMain(...).

The reason why I believe that[defer block](https://github.com/kubernetes-sigs/e2e-framework/blob/332536f9ed7ed22fd069c5622b057a91ceeb1208/pkg/env/env.go#L353-L377) needs to be called is that's where the finish steps are run in case an error happens during a setup step.

I hope despite my lack of brevity, I was able to show my limited understanding, and perhaps you or another person might help me figure out a better way to do the testing.